pkg/apis/pipeline/v1alpha1/volume_resource.go

```diff
@@ -0,0 +1,188 @@
+func (s *VolumeResource) SetSourceDirectory(srcDir string) { s.SourceDir = srcDir }
+
+// GetUploadContainerSpec gets container spec for gcs resource to be uploaded like
+// set environment variable from secret params and set volume mounts for those secrets
+func (s *VolumeResource) GetUploadContainerSpec(sourcePath string) ([]corev1.Container, error) {
+	return []corev1.Container{{
+		Name:    names.SimpleNameGenerator.RestrictLengthWithRandomSuffix(fmt.Sprintf("upload-mkdir-%s", s.Name)),
+		Image:   *BashNoopImage,
+		Command: []string{"/ko-app/bash"},
+		Args: []string{
+			"-args", strings.Join([]string{"mkdir", "-p", filepath.Join(VolumeMountDir, sourcePath)}, " "),
+		},
+		VolumeMounts: []corev1.VolumeMount{s.GetPvcMount()},
+	}, {
+		Name:    names.SimpleNameGenerator.RestrictLengthWithRandomSuffix(fmt.Sprintf("upload-copy-%s", s.Name)),
+		Image:   *BashNoopImage,
+		Command: []string{"/ko-app/bash"},
+		Args: []string{
+			"-args", strings.Join([]string{"cp", "-r", fmt.Sprintf("%s/.", s.SourceDir), filepath.Join(VolumeMountDir, sourcePath)}, " "),
+		},
+		VolumeMounts: []corev1.VolumeMount{s.GetPvcMount()},
+	}}, nil
+}
+
+// GetDownloadContainerSpec returns an array of container specs to download gcs storage object
+func (s *VolumeResource) GetDownloadContainerSpec(sourcePath string) ([]corev1.Container, error) {
+
+	return []corev1.Container{
+		CreateDirContainer(s.Name, sourcePath), {
+			Name:    names.SimpleNameGenerator.RestrictLengthWithRandomSuffix(fmt.Sprintf("download-copy-%s", s.Name)),
+			Image:   *BashNoopImage,
+			Command: []string{"/ko-app/bash"},
+			Args: []string{
+				"-args", strings.Join([]string{"cp", "-r", fmt.Sprintf("%s/.", filepath.Join(VolumeMountDir, s.SourceDir)), sourcePath}, " "),
+			},
+			VolumeMounts: []corev1.VolumeMount{s.GetPvcMount()},
+		}}, nil
+}
+
+// GetPVCSpec returns the PVC to create for a given TaskRun
+func (s *VolumeResource) GetPVCSpec(tr *TaskRun, pvcSize resource.Quantity) *corev1.PersistentVolumeClaim {
+	return &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:       tr.Namespace,
+			Name:            s.Name,
+			OwnerReferences: tr.GetOwnerReference(),
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+			Resources: corev1.ResourceRequirements{
+				Requests: map[corev1.ResourceName]resource.Quantity{
+					corev1.ResourceStorage: pvcSize,
+				},
+			},
+		},
+	}
+}
+
+// GetPvcMount returns a mounting of the volume with the mount path /volumeresource
+func (s *VolumeResource) GetPvcMount() corev1.VolumeMount {
+	return corev1.VolumeMount{
+		Name:      s.Name,         // resource pvc name
+		MountPath: VolumeMountDir, // nothing should be mounted here
+	}
+}
+
+func (s *VolumeResource) GetDownloadVolumeSpec(ts *TaskSpec, tr *TaskRun, c kubernetes.Interface) ([]corev1.Volume, error) {
```

Comment:
can has docstrings for all new functions 🙏

Reply:
do you think it could make sense to pass specifically in the function for creating volumes, instead of the whole `kubernetes.Interface`? (would probably make it easier to write some unit tests for that also)